new Numeric field component

helloWorldCindy opened the pull request:
**This project uses [semver](http://semver.org), please check the scope of this pr:**

- [ ] #none# - documentation fixes and/or test additions
- [ ] #patch# - backwards-compatible bug fix
- [x] #minor# - adding functionality in a backwards-compatible manner
- [ ] #major# - incompatible API change

# CHANGELOG

Add numeric selection component

Commit:
created numveric selection field component

Commit:
add disabled style and get rid of semicolons, change error style

sglanzer-fire commented:
So before I start the review I just wanted to give you a heads up - don't be worried if there is a volume of comments - one of Carl's latest reviews had around 90+, it's simply and educational thing and we tend to be picky since we're productizing components for multiple apps to consume.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+import Ember from 'ember'
+import {PropTypes} from 'ember-prop-types'
+import computed, {readOnly} from 'ember-computed-decorators'
+const {isPresent} = Ember
```
We've been moving towards a convention recently of grouping external (from other packages) and internal (from the same package) imports - which would look like

```
import Ember from 'ember'
const {isPresent} = Ember
import computed, {readOnly} from 'ember-computed-decorators'
import {PropTypes} from 'ember-prop-types'

import Component from './frost-component'
import layout from '../templates/components/frost-numeric-select' 
```

Also note the alphabetical ordering and putting the destructuring of an import directly beside the import

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+/**
+ * Component definition for the numeric selection field component
```
I'm going to suggest naming this component something other than a `-select` since there is a very specific connotation for that word.  How about `frost-numeric-field`?

helloWorldCindy replied:
Fixed. The component is named 'frost-numeric-field' now.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    tabindex: PropTypes.number,
+    type: PropTypes.string,
+    isArrowVisible: PropTypes.bool,
+    isAllowInputNegative: PropTypes.bool,
```
How about `allowNegativeValues` instead of the slightly awkward `isAllow`?

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    type: PropTypes.string,
+    isArrowVisible: PropTypes.bool,
+    isAllowInputNegative: PropTypes.bool,
+    isAllowUseDownButtonToNegative: PropTypes.bool,
```
Can we calculate this off the `isAllowInputNegative`?  Seems like it would be possible to remove this from the interface

dafortin replied:
I will expect this to be done internally

helloWorldCindy replied:
So when the user is allowed to input negative number, they can also use the down button to get to the negative number?  And when the user is not allowed to input negative number, he can't use the down button to get them?

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    required: PropTypes.bool,
+    selectionDirection: PropTypes.string,
+    spellcheck: PropTypes.bool,
+    value: PropTypes.string,
```
I believe the `value` should be required

sglanzer-fire replied:
Ember (and other frameworks) promote a concept known in Ember as Data Down, Actions Up https://dockyard.com/blog/2016/11/18/checkbox-list-ember

Basically the idea is that the state required for a component to function should be held outside of the component and component events should fire to that layer to mutate the state.

This means that `value` is your Data Down and you need a callback like `onInput` to act as the Action Up.

A consumer would have a property that would store the actual state, something like 

https://github.com/ciena-frost/ember-frost-core/blob/master/tests/dummy/app/pods/toggle/template.hbs#L12
and
https://github.com/ciena-frost/ember-frost-core/blob/master/tests/dummy/app/pods/toggle/controller.js#L23

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+      isAllowInputNegative: true,
+      isAllowUseDownButtonToNegative:false,
+      showError: false,
+      errorMessage: 'Cannot input invalid character',
```
Maybe `Input must be numeric`?

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  // == Functions =============================================================
+
+  _isNumberKey (e) {
```
`core` has a keycode enum that would be good to use here https://github.com/ciena-frost/ember-frost-core/blob/master/addon/utils/key-codes.js

helloWorldCindy replied:
Can I use those key code in onKeyPress function? Seems like the key code in the event params is ASCII code.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    let decimalExist = this.get('value') && this.get('value').indexOf('.') >=0
+    let charCode = (e.which) ? e.which : e.keyCode
+
+    if(!this.get('isAllowInputNegative')){
```
Minor, but for clarity I would prefer to check the positive condition (without `!`) and put the negative condition in the `else`

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+   */
+  @computed('isArrowVisible')
+  class (isArrowVisible) {
+    return isArrowVisible ? 'frost-numeric-select-button' : 'frost-numeric-select'
```
I'd like this to be a conditional modifier on the base class, so something like `.frost-numeric-select` as the base class and `frost-numeric-select.arrow-visible` as the modifier, where the `arrow-visible` would provide the right padding, etc.

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    receivedHook: PropTypes.string,
+    tabindex: PropTypes.number,
+    type: PropTypes.string,
+    isArrowVisible: PropTypes.bool,
```
How about `isIncrementControlVisible`?

helloWorldCindy replied:
Fixed. `isArrowVisible` prop has been already changed to `isIncrementControlVisible`.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+          event.preventDefault();
+      }
+      else {
+        this.set('showError', false)
```
I don't think we can clear off the error if an invalid character is still present.  How about calculating the numeric error state using a computed property based on the `value`?

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+    _onInput (event) {
+      if (isPresent(this.get('_eventProxy.input'))) {
+        // Add id and value for legacy support
```
We don't need the legacy support here, so we can just send the event as-is

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+        event.value = event.target.value
+        this._eventProxy.input(event)
+      }
+    },
```
Really minor, but a linebreak is good between properties (and header comments)

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+      }
+    },
+    _onPlusClick (event) {
+      if (this.$('input').val()){
```
Do we need the `$('input').val()` here?  Wouldn't `value` be a better choice?

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+        if (!this.get('isAllowUseDownButtonToNegative')) {
+          if(value >= 0)
+            this._setCalculatedValue(value,numberOfDigit)
```
This is where I would replace the `set` with an `onInput` callback

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+        let numberOfDigit = this._calculateDigit(this.$('input').val())
+        var value = parseFloat(this.get('value'))+1
+
+        if (!this.get('isAllowUseDownButtonToNegative')) {
```
Shouldn't be necessary for the `plus` right?

helloWorldCindy replied:
Fixed. This condition is now removed.

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+$arrow-focus-color: $frost-color-blue-1 !global;
+$error-text-color: $frost-color-input-error-border;
+$disabled-border: 1px solid $frost-color-input-disabled-border;
+$disabled-bg: $frost-color-input-disabled-bg;
```
I'd like to see these variables prefixed since they're going to end up in the global namespace - so

`$frost-number-select-height: 30px;` etc.

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+    border: $focus-border;
+  }
+}
+.error {
```
Be careful with generic classes - CSS is a global namespace so this could inadvertently apply to elements you wouldn't have expect to impact.  It's much safer to make this `frost-numeric-select-error`

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+  font-size: 13px;
+  color: $error-text-color;
+}
+.container{
```
Prefix

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+  right: 0px;
+  top: 0.7px;
+}
+.minusButton {
```
Prefix

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+    }
+  }
+}
+.plusButton {
```
There likely isn't a big different between `minus` and `plus` CSS, so I'd recommend writing it like

```
.minusButton,
.plusButton {
  // Common CSS
}

.minusButton {
  // Specific CSS
}

.plusButton {
  // Specific CSS
}
```

helloWorldCindy replied:
Fixed.

job13er replied:
Also, aren't we using `kebab-case` for class names? So shouldn't it be `minus-button`, `plus-button`? The sass lint should actually complain about that (IIRC).

job13er replied:
Looks like it does: 
```
  81:2  error  Class '.plusButton' should be written in lowercase with hyphens  class-name-format
```
Also, you have a lot of js lint and sass lint errors, which is why the build is failing. You can run: 
```
$ lint-all-the-things
```
at the root of the `ember-frost-core` repo to see the list of all the errors. Ideally, you'll want to configure your editor to show linting errors, then you can just open the files referenced in the results of `lint-all-the-things` and fix the issues your editor shows in each one. Particularly for property order linting of sass, not having it displayed in your editor will make it much more painful to fix.

sglanzer-fire replied:
Yeah, we're using kebab-case

sglanzer-fire replied:
Cindy didn't have sass-linting in her editor, it's on now

sglanzer-fire commented on addon/styles/_frost-numeric-select.scss:
```diff
@@ -0,0 +1,99 @@
+  font-size: 14px;
+  height: 14px;
+  width: 25px;
+  padding: 0px 6.2px 0px 6px;
```
UX likes to stick to units of 5 when possible - also, even if they recommend `6px`, I wouldn't use decimal pixels (`6.2`) - it feels overly precise and not all browsers will respect it

helloWorldCindy replied:
Fixed.

sglanzer-fire commented on addon/templates/components/frost-numeric-select.hbs:
```diff
@@ -0,0 +1,77 @@
+      }}
+    {{/if}}
+  </div>
+  {{#if showError}}
```
We _may_ need a generic error - e.g. "Hey, this field needs a value greater than 20" and the numeric error - I'd say the numeric error (`Input must be numeric`) would take precedence, so maybe we can do a computed property `_errorMessage` where the `errorMessage` (external) and the `_numericErrorMessage` (internal) are inputs to the computed property (CP) and we return the `_numericErrorMessage` if it's present.

I think that could remove the need for `showError` and the condition could just be `{{#if _errorMessage}}`

helloWorldCindy replied:
Fixed.

sglanzer-fire reviewed (changes requested):
Good first pass!  I'd also like to see some tests added, if you would like some help with testing @juwara0 is available to mentor

job13er commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+  propTypes: {    // options
+    align: PropTypes.string,
+    isHookEmbedded: PropTypes.bool,
+    receivedHook: PropTypes.string,
```
These two are obsolete now with the base frost component supporting `hookPrefix`, `frost-text` hasn't been updated to use `hookPrefix` yet because it would be a breaking chance, but new components should certainly not use these two.

sglanzer-fire replied:
I wasn't entirely sure on that, probably shouldn't have used `frost-text` as my reference material for thinking about it

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  // == PropTypes =============================================================
+
+  propTypes: {    // options
```
`// options` should be on the next line.

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  _isNumberKey (e) {
+    let decimalExist = this.get('value') && this.get('value').indexOf('.') >=0
+    let charCode = (e.which) ? e.which : e.keyCode
```
We can do something like this `e.which || e.keyCode` if `e.which` is undefined/null you will get the value of `e.keyCode`

helloWorldCindy replied:
Fixed.

job13er replied:
Is 0 a valid key code? Probably not, so urn should be safe, just remember, if it's anything falsy it'll move to the other side of the ||, including `0`, `''`, or `false`. I know we've had at least one bug where a `false` value wasn't being handled properly because of a check like that. Just a reminder to be cautious.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+  // == Functions =============================================================
+
+  _isNumberKey (e) {
+    let decimalExist = this.get('value') && this.get('value').indexOf('.') >=0
```
I will prefer to have a `const value = this.get('value')` instead of calling get twice.

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+    }
+  },
+
+  //calculate how many digit a value has
```
I will recommend using JsDoc format.

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  // == Functions =============================================================
+
+  _isNumberKey (e) {
```
Is this function failing the eslint complexity?

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  //calculate how many digit a value has
+  _calculateDigit (value) {
+    return value.slice(value.indexOf('.')+1).length
```
`value.indexOf('.') + 1`

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  //calculate how many digit a value has
+  _calculateDigit (value) {
+    return value.slice(value.indexOf('.')+1).length
```
Do we want to support only the `.` digit mark? https://en.wikipedia.org/wiki/Decimal_mark

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  //set calculated and properly formated value
+  _setCalculatedValue (value, numberOfDigit) {
+    if(this.$('input').val().indexOf('.') > -1) {
```
If I have the following:
1.00 => 1.00
1.000 => 1.00
1 => 1

Is that what is expected?

helloWorldCindy replied:
This function is used for prevent floating point number precision in JS calculation. (e.g 1.22 - 1 will equals 0.199998, now it will be 1.22 - 1 = 0.22). I am not sure if it is required though.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+
+  //set calculated and properly formated value
+  _setCalculatedValue (value, numberOfDigit) {
+    if(this.$('input').val().indexOf('.') > -1) {
```
We should probably add a function to determine if it's a number with decimals and do this check there instead of copying this check everywhere.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+  // == Lifecycle Hooks =======================================================
+  init () {
+    this._super(...arguments)
+    this.set('receivedHook', this.get('hook'))
```
not necessary as @job13er mentioned.

helloWorldCindy replied:
Fixed. It is removed now.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+      }
+    },
+    _onPlusClick (event) {
+      if (this.$('input').val()){
```
Can you create a variable to store the value?

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+      if (this.$('input').val()){
+        //calculate how many digit a value has
+        let numberOfDigit = this._calculateDigit(this.$('input').val())
+        var value = parseFloat(this.get('value'))+1
```
Why are we adding `+ 1` to the value?

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+      if (this.$('input').val()){
+        //calculate how many digit a value has
+        let numberOfDigit = this._calculateDigit(this.$('input').val())
+        var value = parseFloat(this.get('value'))+1
```
` + 1`

helloWorldCindy replied:
Fixed.

dafortin commented on addon/components/frost-numeric-select.js:
```diff
@@ -0,0 +1,183 @@
+        let numberOfDigit = this._calculateDigit(this.$('input').val())
+        let value = parseFloat(this.get('value'))-1
+
+        if (!this.get('isAllowUseDownButtonToNegative')) {
```
It might be interesting to disable the minus button instead of doing something like this?
We might simply just not decrease if the number is  === 0

helloWorldCindy replied:
Added a computed property to disable button when the number is less or equal to zero. Fixed.

dafortin commented:
Great job :) Only a few small comments.

dafortin commented:
Opps closed the PR by mistake :(

Commit:
Fixed most of the error mentioned in the comment and eslint errors. What's left:
calculate allowUseDownButtonToNegative,
replace set with an onInput callback,
only enable user enter dash sign at index 0 of the value

Commit:
fix eslint error in sass

Commit:
fixed negative mark problem, remove isAllowUseDownButtonToNegative prop

Commit:
fixed onInput function, tests in progress

Commit:
added unit test and integration test

Commit:
add onInput function in dummy app controller

Commit:
replace set with onInput callback function

sglanzer-fire commented on addon/components/frost-numeric-field.js:
```diff
@@ -0,0 +1,232 @@
+    * @param {string} value - value of the input
+    * @param {number} numberOfDigit - number of digit
+ */
+  _setCalculatedValue (value, numberOfDigit) {
```
I'd probably suggest renaming this to reflect that it's not a `set` now, but we can clean that up the next time we're in the code

sglanzer-fire commented:
Approved - the linting is failing your build currently, so you'll need to fix that, good otherwise

[![Approved with PullApprove](https://img.shields.io/badge/pullapprove-approved-brightgreen.svg)](https://pullapprove.com/ciena-frost/ember-frost-core/pull-request/449/?utm_source=github-pr&utm_medium=comment-badge&utm_campaign=ciena-frost/ember-frost-core)

sglanzer-fire commented on tests/dummy/app/pods/numeric-field/controller.js:
```diff
@@ -0,0 +1,13 @@
+
+export default Controller.extend({
+  queryParams: ['value'],
+  value: '3',
```
Sorry, I just noticed this - the value for a numeric field should really be in a numeric format - is the field capable of handling something other than a string (a number)?

sglanzer-fire reviewed (dismissed):
I just need confirmation that this can handle numeric values rather than string values

rox163 commented:
I'm guessing this is still a component we want? @job13er @helloWorldCindy @dafortin  @juwara0 
Would like to see some movement on this PR accordingly.

job13er commented:
@rox163 I honestly don't know. I don't know the use-case it's trying to solve. I personally don't have any need of a "numeric selection component" although that description doesn't really say much about what it is and why it's needed/wanted. I'm guessing it's like `frost-select` but where the values can be numbers and not just strings? I've always just used `parseInt()` and a regular `frost-select` when working with a select that was choosing a number.

juwara0 commented:
@rox163 @job13er - I found this past PR https://github.com/ciena-frost/ember-frost-core/pull/376 that was closed in favor of implementing a separate component. There is a little more information in the comments. I found this issue: https://github.com/ciena-frost/ember-frost-core/issues/405 that @sglanzer opened to track the task.

job13er commented:
@juwara0 Thanks for the context. I'd remove any reference to "select" in the PR description and changelog then, as it's completely misleading. The component name is `frost-numeric-field` which I'm not sure how I feel about, since it's basically `frost-text` but for a number, would `frost-number` be more appropriate? Naming aside, I have no problem with the component being created. I've just used `frost-text` for numeric inputs thus far.

rox163 commented:
@helloWorldCindy could you please follow up with any remaining comments in the PR and also address the comments here https://github.com/ciena-frost/ember-frost-core/issues/405 and push updates please? Thanks

helloWorldCindy commented:
@rox163 Sure, will do it ASAP

Commit:
Merge branch 'master' into numericSelect

Commit:
make input field support input number

notmessenger commented on addon/components/frost-numeric-field.js:
```diff
@@ -184,6 +187,7 @@ export default Component.extend(FrostEventsProxyMixin, {
 
   // == Lifecycle Hooks =======================================================
   init () {
+    // this.set('value', String(this.get('value')))
```
Remove

helloWorldCindy replied:
Fixed.

Commit:
remove useless code

Commit:
fix tests

Commit:
fix eslint errors in tests

Commit:
fix spellcheck test failing error

Commit:
Merge branch 'master' into numericSelect

Commit:
Merge branch 'master' into numericSelect

Commit:
fix tests to increase code coverage

Commit:
fix eslint error

Commit:
fix tests cases to increase tests coverage, also change input type from number to string

Commit:
add unit tests

Commit:
fix eslint error

Commit:
Merge branch 'master' into numericSelect

Commit:
fixing tests to increase test coverage

NickLewanowicz reviewed (approved):
Great work Cindy, Amazon should be proud!
